_posts/2021-05-18-managing-react-application-state-with-context.md

```diff
@@ -0,0 +1,177 @@
+
+const CountProvider: FC<{ init?: Partial<State> }> = ({ children, init }) => {
+  const initialState: State = {
+    count: 0,
+    ...init,
+  };
+
+  const [state, dispatch] = useReducer(CountReducer, initialState);
+
+  return (
+    <CountStateContext.Provider value={state}>
+      <CountDispatchContext.Provider value={dispatch}>
+        {children}
+      </CountDispatchContext.Provider>
+    </CountStateContext.Provider>
+  );
+};
+```
+
+I then wrap our App in the `CountProvider` component, ensuring that all of the components that rely on the `State` or will perform `Action`s on that state are encapsulated as children of this component.
+
+```javascript
+// App.tsx
+function App() {
+  return (
+    <CountProvider>
+      <CountDisplay />
+      <CountButtons />
+      <CountInput />
+    </CountProvider>
+  );
+}
+```
+
+Finally, I provide some convenience functions which allow us to easily access the our `State` and `Dispatch` actions within components along with providing some error feedback if we are trying to access them outside of our `CountProvider`.
+
+```javascript
+// countProvider.tsx
+export function useCount(): [State, Dispatch] {
+  const state = useContext(CountStateContext);
+  const dispatch = useContext(CountDispatchContext);
+  if (state === undefined || dispatch === undefined) {
+    throw new Error("useCount must be used within a CountProvider");
+  }
+  return [state, dispatch];
+}
+
+export function increment(dispatch: Dispatch) {
+  dispatch({ type: "increment" })
+}
+
+export function decrement(dispatch: Dispatch) {
+  dispatch({ type: "decrement" })
+}
+
+export function setValue(dispatch: Dispatch, value: number) {
+  dispatch({ type: "setValue", payload: { value } });
+}
+```
+
+I've included some examples below of what this looks like within our components.
+
+```javascript
+// CountDisplay.tsx
+const CountDisplay = () => {
+  const [{ count }] = useCount();
+
+  return <div>{count}</div>;
+};
+
+// CountButtons.tsx
+const CountButtons = () => {
+  const [, dispatch] = useCount();
+
+  const onIncrement = () => {
+    increment(dispatch);
+  };
+
+  const onDecrement = () => {
+    decrement(dispatch);
+  };
+
+  return (
+    <div>
+      <button onClick={onIncrement}>Increment</button>
+      <button onClick={onDecrement}>Decrement</button>
+    </div>
+  );
+};
+```
+
+And thats how we are able to manage application state using Context which is built into React. 
+
+There is one thing that should be mentioned and that is *performance*. Whenever the value of the Context changes, it triggers a re-render of all the child components of that Context which, in our use case isn't all that bad. However, if you have a large number of child components or those components are computationally expensive to render, you may want to look at the solutions provided in this [Github Issue](https://github.com/facebook/react/issues/15156#issuecomment-474590693). 
+
```

Comment:
Nice sign off. I'd also include a hook to get in touch with us too:

https://www.mechanicalrock.io/lets-get-started/

Reply:
Before signing off, I think you should provide examples on how to test the Context because I know people are confused how to do it. I appreciate there are many ways to test it, but by defining it here you'll give noobs an end-to-end blog.

Reply:
Again, purely because I want to avoid making it too long, could be a follow up however there is also an example of how to test it in the example repo which I have linked.